wireguard/metrics.go

```diff
@@ -0,0 +1,227 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/projectcalico/felix/netlinkshim"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+const (
+	labelHostname      = "hostname"
+	labelPublicKey     = "public_key"
+	labelInterfaceName = "iface"
+	labelListenPort    = "listen_port"
+	labelPeerKey       = "peer_key"
+	labelPeerEndpoint  = "peer_endpoint"
+)
+
+var _ prometheus.Collector = (*Metrics)(nil)
+
+const (
+	wireguardMetaFQName   = "wireguard_meta"
+	wireguardMetaHelpText = "wireguard interface and runtime metadata"
+
+	wireguardLatestHandshakeIntervalFQName   = "wireguard_latest_handshake_seconds"
+	wireguardLatestHandshakeIntervalHelpText = "wireguard interface latest handshake unix timestamp in seconds to a peer"
+
+	wireguardBytesSentFQName   = "wireguard_bytes_sent_total"
+	wireguardBytesSentHelpText = "wireguard interface total outgoing bytes to peer"
+
+	wireguardBytesRcvdFQName   = "wireguard_bytes_rcvd_total"
+	wireguardBytesRcvdHelpText = "wireguard interface total incoming bytes to peer"
+)
+
+func init() {
+	prometheus.MustRegister(
+		MustNewWireguardMetrics(),
+	)
+}
+
+type Metrics struct {
+	hostname           string
+	newWireguardClient func() (netlinkshim.Wireguard, error)
+	logCtx             *logrus.Entry
+
+	peerRx, peerTx map[wgtypes.Key]int64
+}
+
+func (collector *Metrics) Describe(chan<- *prometheus.Desc) {}
+
+func (collector *Metrics) Collect(m chan<- prometheus.Metric) {
+	collector.refreshStats(m)
+}
+
+func MustNewWireguardMetrics() *Metrics {
+	wg, err := NewWireguardMetrics()
+	if err != nil {
+		logrus.Panic(err)
+	}
+	return wg
+}
+
+func NewWireguardMetrics() (*Metrics, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		logrus.WithError(err).Error("cannot register wireguard metrics stats")
+		return nil, err
+	}
+	return NewWireguardMetricsWithShims(hostname, netlinkshim.NewRealWireguard), nil
+}
+
+func NewWireguardMetricsWithShims(hostname string, newWireguardClient func() (netlinkshim.Wireguard, error)) *Metrics {
+	logrus.WithField("hostname", hostname).Debug("created wireguard collector for host")
+	return &Metrics{
+		hostname:           hostname,
+		newWireguardClient: newWireguardClient,
+		logCtx: logrus.WithFields(logrus.Fields{
+			"prometheus_collector": "wireguard",
+		}),
+
+		peerRx: map[wgtypes.Key]int64{},
+		peerTx: map[wgtypes.Key]int64{},
+	}
+}
+
+func (collector *Metrics) refreshStats(m chan<- prometheus.Metric) {
+	wgClient, err := collector.newWireguardClient()
+	if err != nil {
+		collector.logCtx.WithError(err).Error("error initializing wireguard client devices")
+		return
+	}
+
+	devices, err := wgClient.Devices()
```

Comment:
Do we ever have >1 wireguard tunnel device?  I thought we only had one (so you could just poll that one instead of _all_ tunnels).  Probably not worth the extra effort to make this code _less_ generic but I think it's 50:50 whether a user will want us to report non-Calico Wireguard tunnels.

Reply:
no not really.. we only have wireguard.cali (perhaps 90+% of the time). yeah I wanted to make this initial code as generic as possible.. let me know if you're happy to leave it at its generic state or make it wireguard.cali only or whatever it's set in the felix config

Reply:
OK to leave as generic, I think